spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/ReactiveDatastoreTemplate.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.StreamSupport;
+
+import com.google.cloud.datastore.BaseEntity;
+import com.google.cloud.datastore.BaseKey;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.EntityQuery;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.domain.Example;
+import org.springframework.util.Assert;
+
+/**
+ * A reactive template for Cloud Datastore operations.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.2
+ */
+public class ReactiveDatastoreTemplate {
+
+	private final DatastoreTemplate datastoreTemplate;
+
+	/**
+	 * Constructor.
+	 * @param datastoreTemplate A standard synchronous Datastore Template used to resolve
+	 *     relationships.
+	 */
+	public ReactiveDatastoreTemplate(DatastoreTemplate datastoreTemplate) {
+		Assert.notNull(datastoreTemplate,
+				"A non-null DatastoreTemplate is required.");
+		this.datastoreTemplate = datastoreTemplate;
+	}
+
+	public <T> Mono<T> findById(Object id, Class<T> entityClass) {
+		return performFindByKey(Collections.singleton(id), entityClass).next();
```

Comment:
regarding event, for `mono` results, emitting result when action happens makes sense.